Source/Core/DiscIO/DiscExtracter.cpp

```diff
@@ -0,0 +1,174 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DiscIO/DiscExtracter.h"
+
+#include <algorithm>
+#include <cinttypes>
+#include <optional>
+
+#include "Common/CommonTypes.h"
+#include "DiscIO/Enums.h"
+#include "DiscIO/Filesystem.h"
+#include "DiscIO/Volume.h"
+
+namespace DiscIO
+{
+u64 ReadFile(const Volume& volume, const Partition& partition, const FileInfo* file_info,
+             u8* buffer, u64 max_buffer_size, u64 offset_in_file)
+{
+  if (!file_info || file_info->IsDirectory())
+    return 0;
+
+  if (offset_in_file >= file_info->GetSize())
+    return 0;
+
+  u64 read_length = std::min(max_buffer_size, file_info->GetSize() - offset_in_file);
+
+  DEBUG_LOG(DISCIO, "Reading %" PRIx64 " bytes at %" PRIx64 " from file %s. Offset: %" PRIx64
+                    " Size: %" PRIx32,
+            read_length, offset_in_file, file_info->GetPath().c_str(), file_info->GetOffset(),
+            file_info->GetSize());
+
+  volume.Read(file_info->GetOffset() + offset_in_file, read_length, buffer, partition);
+
+  return read_length;
+}
+
+bool ExportFile(const Volume& volume, const Partition& partition, const FileInfo* file_info,
+                const std::string& export_filename)
+{
+  if (!file_info || file_info->IsDirectory())
+    return false;
+
+  u64 remaining_size = file_info->GetSize();
+  u64 file_offset = file_info->GetOffset();
+
+  File::IOFile f(export_filename, "wb");
+  if (!f)
+    return false;
+
+  bool result = true;
+
+  while (remaining_size)
+  {
+    // Limit read size to 128 MB
+    size_t read_size = (size_t)std::min(remaining_size, (u64)0x08000000);
+
+    std::vector<u8> buffer(read_size);
+
+    result = volume.Read(file_offset, read_size, &buffer[0], partition);
+
+    if (!result)
+      break;
+
+    f.WriteBytes(&buffer[0], read_size);
+
+    remaining_size -= read_size;
+    file_offset += read_size;
+  }
+
+  return result;
+}
+
+bool ExportApploader(const Volume& volume, const Partition& partition,
+                     const std::string& export_folder)
+{
+  if (!IsDisc(volume.GetVolumeType()))
+    return false;
+
+  std::optional<u32> apploader_size = volume.ReadSwapped<u32>(0x2440 + 0x14, partition);
+  const std::optional<u32> trailer_size = volume.ReadSwapped<u32>(0x2440 + 0x18, partition);
```

Comment:
That's a shame because these magic numbers mean nothing to someone who isn't familiar with the code.  Would it make sense for readability to capture them in variables for now?